"Provider io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider not found"

### Describe the bug

I'm trying to integrate OpenTelemetry in a Quarkus application (Quarkus 3.20.0, Java 21), to export TRACES via OTLP, using "http/protobuf" as protocol. When I run the application in pure-java mode, even in dev-mode, it works fine. But when I build a native executable and then I try to run it, I get the following message:
``` 
RuntimeException: Failed to start quarkus 
... 
Caused by: java.util.ServiceConfigurationError: io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider: Provider io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider not found
```
My pom.xml:
```
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-opentelemetry</artifactId>
</dependency>
<dependency>
  <groupId>io.opentelemetry</groupId>
  <artifactId>opentelemetry-exporter-otlp</artifactId>
</dependency>
```
My application.properties:
```
quarkus.otel.traces.enabled=true
quarkus.otel.traces.exporter=otlp
quarkus.otel.metrics.enabled=false
quarkus.otel.metrics.exporter=none
quarkus.otel.logs.enabled=false
quarkus.otel.logs.exporter=none
```
And some environment variables for runtime:
```
QUARKUS_OTEL_SDK_DISABLED=false
QUARKUS_OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
QUARKUS_OTEL_EXPORTER_OTLP_ENDPOINT=...
QUARKUS_OTEL_EXPORTER_OTLP_HEADERS="Authorization=Api-Token ..."
```
As I said, in dev-mode works fine, so it's not a dependency problem.

I also tried to register some classes for reflection using a reflect-config.json file:
```
[
  {
    "name": "io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider",
  },
  {
    "name": "io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter"
  }
]
```
The same with a jni-config.json file, and also tried registering services for io.opentelemetry.sdk.autoconfigure.spi.internal.AutoConfigureListener and io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider.

I even tried some "dummy-loading" of the conflicting provider, by creating an instance of the class and then invoking a method on it:
```
OtlpSpanExporterProvider dummy = new OtlpSpanExporterProvider();
System.out.println("dummy.getName(): "+dummy.getName());
```
However, when I try to build the native application with that code, I get another message: 
```
Unsupported type io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider is reachable.
```
In dev-mode that code works perfectly, and the System.out prints `dummy.getName(): otlp`.


### Expected behavior

The native aplication is build and can be started. When any endpoint is invoked, some traces are exported.

### Actual behavior

Without the last part (the instantiation of OtlpSpanExporterProvider), the native application is generated, but when I try to execute it the aforementioned exception is shown ("Provider io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider not found").
Whit the last part, the native can not be built, and the message "Unsupported type io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider is reachable." is shown.

### How to Reproduce?

1. Create a Quarkus project (v 3.20.0).
2. Include dependency with OpenTelemetry (io.quarkus:quarkus-opentelemetry):
```
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-opentelemetry</artifactId>
        </dependency>
        <dependency>
          <groupId>io.opentelemetry</groupId>
          <artifactId>opentelemetry-exporter-otlp</artifactId>
        </dependency>
        <dependency>
            <groupId>io.opentelemetry</groupId>
            <artifactId>opentelemetry-exporter-sender-okhttp</artifactId>
            <version>1.44.1</version>
        </dependency>
```
3. Configure the app at build time to export traces via otlp (application.properties):
```
quarkus.otel.traces.enabled=true
quarkus.otel.traces.exporter=otlp
quarkus.otel.metrics.enabled=false
quarkus.otel.metrics.exporter=none
quarkus.otel.logs.enabled=false
quarkus.otel.logs.exporter=none
```
4. Build a native executable (JDK 21):
```
$> ./mvnw clean package -Dnative -Dquarkus.native.container-build=true -Dquarkus.native.container-runtime=podman
```
(I use a settings.xml to point to an empty .m2 each time)
5. Set up environment variables:
```
export QUARKUS_OTEL_SDK_DISABLED=false
export QUARKUS_OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
export QUARKUS_OTEL_EXPORTER_OTLP_ENDPOINT=...
export QUARKUS_OTEL_EXPORTER_OTLP_HEADERS="Authorization=Api-Token ..."
```
6. Try to run the application:
$> ./target/opentelemetry-test-0.1.0-runner
```
2025-06-17 16:18:04,292 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:121)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:77)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:48)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:137)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.util.ServiceConfigurationError: io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider: Provider io.opentelemetry.exporter.otlp.internal.OtlpSpanExporterProvider not found
	at java.base@21.0.7/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base@21.0.7/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1219)
	at java.base@21.0.7/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
	at java.base@21.0.7/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base@21.0.7/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base@21.0.7/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at io.opentelemetry.sdk.autoconfigure.internal.SpiHelper.load(SpiHelper.java:162)
	at io.opentelemetry.sdk.autoconfigure.internal.SpiHelper.loadConfigurable(SpiHelper.java:80)
	at io.opentelemetry.sdk.autoconfigure.SpanExporterConfiguration.spanExporterSpiManager(SpanExporterConfiguration.java:82)
	at io.opentelemetry.sdk.autoconfigure.SpanExporterConfiguration.configureSpanExporters(SpanExporterConfiguration.java:64)
	at io.opentelemetry.sdk.autoconfigure.TracerProviderConfiguration.configureTracerProvider(TracerProviderConfiguration.java:58)
	at io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder.build(AutoConfiguredOpenTelemetrySdkBuilder.java:467)
	at io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder$1.apply(OpenTelemetryRecorder.java:87)
	at io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder$1.apply(OpenTelemetryRecorder.java:60)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.createSynthetic(Unknown Source)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.doCreate(Unknown Source)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.create(Unknown Source)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.ComputingCacheContextInstances.computeIfAbsent(ComputingCacheContextInstances.java:19)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.get(Unknown Source)
	at io.opentelemetry.api.OpenTelemetry_Le6zQbzkojAYO_OiKIQWJf4lGa4_Synthetic_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.InstanceImpl.getBeanInstance(InstanceImpl.java:325)
	at io.quarkus.arc.impl.InstanceImpl.getInternal(InstanceImpl.java:309)
	at io.quarkus.arc.impl.InstanceImpl.get(InstanceImpl.java:190)
	at io.quarkus.arc.runtime.BeanContainerImpl.beanInstance(BeanContainerImpl.java:28)
	at io.quarkus.opentelemetry.runtime.tracing.intrumentation.InstrumentationRecorder.setupVertxTracer(InstrumentationRecorder.java:59)
	at io.quarkus.runner.recorded.OpenTelemetryProcessor$setupVertx1012958309.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.OpenTelemetryProcessor$setupVertx1012958309.deploy(Unknown Source)
	... 7 more
```

### Output of `uname -a` or `ver`

Linux uy-pc-0123 6.8.0-60-generic #63-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 15 19:04:15 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21.0.1" 2023-10-17 OpenJDK Runtime Environment (build 21.0.1+12-29) OpenJDK 64-Bit Server VM (build 21.0.1+12-29, mixed mode, sharing)

### Mandrel or GraalVM version (if different from Java)

GraalVM 23.1.2

### Quarkus version or git rev

Quarkus 3.20.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
$> ./mvnw --version 
Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546) 
Maven home: /home/spio/.m2/wrapper/dists/apache-maven-3.9.5-bin/860fdc54/apache-maven-3.9.5 
Java version: 21.0.1, vendor: Oracle Corporation, runtime: /usr/lib/jvm/jdk-21.0.1 
Default locale: en_US, platform encoding: UTF-8 
OS name: "linux", version: "6.8.0-60-generic", arch: "amd64", family: "unix"
```
### Additional information

Simple test project:
[opentelemetry-test.tar.gz](https://github.com/user-attachments/files/20782957/opentelemetry-test.tar.gz)
Steps:
1. Build the native application:
$> ./mvnw clean package -Dnative -Dquarkus.native.container-build=true -Dquarkus.native.container-runtime=podman --settings=./src/main/resources/settings.xml
2. Run the application:
$> ./target/opentelemetry-test-0.1.0-runner

--- a/docs/src/main/asciidoc/opentelemetry.adoc
+++ b/docs/src/main/asciidoc/opentelemetry.adoc
@@ -251,15 +251,89 @@ When creating manual instrumentation, while naming metrics or attributes you sho
 
 == Exporters
 
-=== Default
+=== The Default
 
-The Quarkus OpenTelemetry extension uses its own signal exporters built on top of Vert.x for optimal performance and maintainability.
+The Quarkus OpenTelemetry extension uses its own signal exporters built on top of Vert.x for optimal performance and maintainability. All *Quarkus built in exporters use the OTLP protocol* through a couple of data senders, using `grpc` (the default) and `http/protobuf`.
 
-The exporter is automatically wired with CDI, that's why the `quarkus.otel.traces.exporter` and `quarkus.otel.metrics.exporter` properties default to `cdi`.
+The active exporter is automatically wired by CDI, that's why the `quarkus.otel.traces.exporter`, `quarkus.otel.metrics.exporter` and `quarkus.otel.logs.exporter` properties default value is `cdi`. This is not because of the protocol being used in the data transfer but because of how the exporters are wired.
 
-The `quarkus.otel.exporter.otlp.protocol` defaults to `grpc` but `http/protobuf` can also be used.
+CDI (Context Dependency Injection) will manage the exporters to use, according to the selected protocol or when applications implement their own CDI exporter, like in tests.
+
+The `quarkus.otel.exporter.otlp.protocol` property instructs Quarkus to switch the senders and defaults to `grpc` but `http/protobuf` can also be used.
 
 NOTE: If you change the protocol, you also need to change the port in the endpoint. The default port for `grpc` is `4317` and for `http/protobuf` is `4318`.
+
+=== Using CDI to produce a test exporter
+
+Leaving the default as CDI is particularly useful for tests. In the following example a Span exporter class is wired with CDI and then the telemetry can be used in test code.
+
+Creating a custom `SpanExporter` bean:
+
+[source,java]
+----
+    @ApplicationScoped
+    static class InMemorySpanExporterProducer {
+        @Produces
+        @Singleton
+        InMemorySpanExporter inMemorySpanExporter() {
+            return InMemorySpanExporter.create();
+        }
+    }
+----
+
+Where `InMemorySpanExporter` is a class from the OpenTelemetry test utilities dependency:
+
+[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
+.pom.xml
+----
+    <dependency>
+        <groupId>io.opentelemetry</groupId>
+        <artifactId>opentelemetry-sdk-testing</artifactId>
+        <scope>test</scope>
+    </dependency>
+----
+
+[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
+.build.gradle
+----
+implementation("io.opentelemetry:opentelemetry-sdk-testing")
+----
+
+The bean of that class can be injected to access the telemetry data. This is an example to obtain the spans:
+
+[source, java]
+----
+    @Inject
+    InMemorySpanExporter inMemorySpanExporter;
+
+    //...
+
+    List<SpanData> finishedSpanItems = inMemorySpanExporter.getFinishedSpanItems();
+----
+
+If this is used in an integration test, you should access the class from inside the running process and not from the test class.
+A viable option could be to expose that data through a REST endpoint method:
+
+[source,java]
+----
+    @GET
+    @Path("/export")
+    public List<SpanData> exportTraces() {
+        return inMemorySpanExporter.getFinishedSpanItems()
+                .stream()
+                .filter(sd -> !sd.getName().contains("export")) <1>
+                .collect(Collectors.toList());
+    }
+----
+<1> This excludes calls to the export endpoint itself.
+
+For more details please take a look at the https://github.com/quarkusio/quarkus/blob/main/integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java[ExporterResource] in the Quarkus integration tests.
+
+=== The OpenTelemetry OTLP exporter
+
+This is currently not supported in Quarkus. Configuration example for traces: `quarkus.otel.tracing.exporter=otlp`.
+
+However, it's also not needed because Quarkus own default exporters will send data using the OTLP protocol.
 
 === On Quarkiverse
 Additional exporters will be available in the Quarkiverse https://docs.quarkiverse.io/quarkus-opentelemetry-exporter/dev/index.html[quarkus-opentelemetry-exporter] project.

--- a/extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java
+++ b/extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java
@@ -1,6 +1,7 @@
 package io.quarkus.opentelemetry.deployment.exporter.otlp;
 
 import static io.quarkus.opentelemetry.runtime.config.build.ExporterType.Constants.CDI_VALUE;
+import static io.quarkus.opentelemetry.runtime.config.build.ExporterType.Constants.OTLP_VALUE;
 
 import java.util.List;
 import java.util.function.BooleanSupplier;
@@ -9,6 +10,7 @@ import java.util.logging.Level;
 import jakarta.enterprise.inject.Instance;
 import jakarta.inject.Singleton;
 
+import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.ClassType;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.ParameterizedType;
@@ -20,8 +22,10 @@ import io.opentelemetry.sdk.trace.SpanProcessor;
 import io.opentelemetry.sdk.trace.export.SpanExporter;
 import io.quarkus.arc.deployment.BeanDiscoveryFinishedBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
+import io.quarkus.arc.deployment.ValidationPhaseBuildItem;
 import io.quarkus.deployment.annotations.*;
 import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.LogCategoryBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigBuilderBuildItem;
 import io.quarkus.opentelemetry.runtime.config.build.OTelBuildConfig;
@@ -29,6 +33,7 @@ import io.quarkus.opentelemetry.runtime.config.build.exporter.OtlpExporterBuildC
 import io.quarkus.opentelemetry.runtime.config.runtime.exporter.OtlpExporterConfigBuilder;
 import io.quarkus.opentelemetry.runtime.exporter.otlp.OTelExporterRecorder;
 import io.quarkus.opentelemetry.runtime.exporter.otlp.tracing.LateBoundSpanProcessor;
+import io.quarkus.runtime.configuration.ConfigurationException;
 import io.quarkus.tls.TlsConfigurationRegistry;
 import io.quarkus.tls.deployment.spi.TlsRegistryBuildItem;
 import io.quarkus.vertx.core.deployment.CoreVertxBuildItem;
@@ -38,6 +43,7 @@ public class OtlpExporterProcessor {
 
     private static final DotName METRIC_EXPORTER = DotName.createSimple(MetricExporter.class.getName());
     private static final DotName LOG_RECORD_EXPORTER = DotName.createSimple(LogRecordExporter.class.getName());
+    private static final DotName OKHTTP_INTERCEPTOR = DotName.createSimple("okhttp3.Interceptor");
 
     static class OtlpTracingExporterEnabled implements BooleanSupplier {
         OtlpExporterBuildConfig exportBuildConfig;
@@ -75,6 +81,34 @@ public class OtlpExporterProcessor {
         }
     }
 
+    static class OtlpUpstreamExporterEnabled implements BooleanSupplier {
+        OtlpExporterBuildConfig exportBuildConfig;
+        OTelBuildConfig otelBuildConfig;
+
+        public boolean getAsBoolean() {
+            return otelBuildConfig.enabled() &&
+                    exportBuildConfig.enabled() &&
+                    (isLogsOtlp(otelBuildConfig) ||
+                            isMetricsOtlp(otelBuildConfig) ||
+                            isTracesOtlp(otelBuildConfig));
+        }
+
+        private boolean isLogsOtlp(OTelBuildConfig otelBuildConfig) {
+            return otelBuildConfig.logs().enabled().orElse(Boolean.TRUE) &&
+                    otelBuildConfig.logs().exporter().contains(OTLP_VALUE);
+        }
+
+        private boolean isMetricsOtlp(OTelBuildConfig otelBuildConfig) {
+            return otelBuildConfig.metrics().enabled().orElse(Boolean.TRUE) &&
+                    otelBuildConfig.metrics().exporter().contains(OTLP_VALUE);
+        }
+
+        private boolean isTracesOtlp(OTelBuildConfig otelBuildConfig) {
+            return otelBuildConfig.traces().enabled().orElse(Boolean.TRUE) &&
+                    otelBuildConfig.traces().exporter().contains(OTLP_VALUE);
+        }
+    }
+
     @BuildStep
     void logging(BuildProducer<LogCategoryBuildItem> log) {
         // Reduce the log level of the exporters because it's too much, and we do log important things ourselves.
@@ -85,6 +119,20 @@ public class OtlpExporterProcessor {
     @BuildStep
     void config(BuildProducer<RunTimeConfigBuilderBuildItem> runTimeConfigBuilderProducer) {
         runTimeConfigBuilderProducer.produce(new RunTimeConfigBuilderBuildItem(OtlpExporterConfigBuilder.class));
+    }
+
+    @BuildStep(onlyIf = OtlpUpstreamExporterEnabled.class)
+    void errorIfOtlpUpstreamIsUsed(CombinedIndexBuildItem indexBuildItem,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> validationErrors) {
+        final ClassInfo okHttpInterceptor = indexBuildItem.getIndex().getClassByName(OKHTTP_INTERCEPTOR);
+
+        // User might be forcing the upstream libraries and we should let them.
+        if (okHttpInterceptor == null) {
+            validationErrors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
+                    new ConfigurationException("OpenTelemetry exporter set to 'otlp' but upstream dependencies not found. " +
+                            "Please notice: The Quarkus default exporters are already OTLP protocol compliant. " +
+                            "For more see: https://quarkus.io/guides/opentelemetry#default-2")));
+        }
     }
 
     @SuppressWarnings("deprecation")

--- a/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamAllExporterErrorTest.java
+++ b/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamAllExporterErrorTest.java
@@ -1,0 +1,40 @@
+package io.quarkus.opentelemetry.deployment.exporter.otlp;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class OtlpUpstreamAllExporterErrorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withEmptyApplication()
+            .overrideConfigKey("quarkus.otel.traces.exporter", "otlp")
+            .overrideConfigKey("quarkus.otel.metrics.enabled", "true")
+            .overrideConfigKey("quarkus.otel.metrics.exporter", "otlp")
+            .overrideConfigKey("quarkus.otel.logs.enabled", "true")
+            .overrideConfigKey("quarkus.otel.logs.exporter", "otlp")
+            .assertException(t -> {
+                assertEquals(DeploymentException.class, t.getClass());
+                Assertions.assertTrue(t.getMessage().contains(
+                        "io.quarkus.runtime.configuration.ConfigurationException: " +
+                                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found"));
+            });
+
+    @Inject
+    OpenTelemetry openTelemetry;
+
+    @Test
+    void testOpenTelemetryButNoBatchSpanProcessor() {
+        Assertions.fail("Test should not be run as deployment should fail with: " +
+                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found... ");
+    }
+}

--- a/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamLogsExporterErrorTest.java
+++ b/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamLogsExporterErrorTest.java
@@ -1,0 +1,37 @@
+package io.quarkus.opentelemetry.deployment.exporter.otlp;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class OtlpUpstreamLogsExporterErrorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withEmptyApplication()
+            .overrideConfigKey("quarkus.otel.logs.enabled", "true")
+            .overrideConfigKey("quarkus.otel.logs.exporter", "otlp")
+            .assertException(t -> {
+                assertEquals(DeploymentException.class, t.getClass());
+                Assertions.assertTrue(t.getMessage().contains(
+                        "io.quarkus.runtime.configuration.ConfigurationException: " +
+                                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found"));
+            });
+
+    @Inject
+    OpenTelemetry openTelemetry;
+
+    @Test
+    void testOpenTelemetryButNoBatchSpanProcessor() {
+        Assertions.fail("Test should not be run as deployment should fail with: " +
+                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found... ");
+    }
+}

--- a/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamMetricsExporterErrorTest.java
+++ b/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamMetricsExporterErrorTest.java
@@ -1,0 +1,37 @@
+package io.quarkus.opentelemetry.deployment.exporter.otlp;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class OtlpUpstreamMetricsExporterErrorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withEmptyApplication()
+            .overrideConfigKey("quarkus.otel.metrics.enabled", "true")
+            .overrideConfigKey("quarkus.otel.metrics.exporter", "otlp")
+            .assertException(t -> {
+                assertEquals(DeploymentException.class, t.getClass());
+                Assertions.assertTrue(t.getMessage().contains(
+                        "io.quarkus.runtime.configuration.ConfigurationException: " +
+                                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found"));
+            });
+
+    @Inject
+    OpenTelemetry openTelemetry;
+
+    @Test
+    void testOpenTelemetryButNoBatchSpanProcessor() {
+        Assertions.fail("Test should not be run as deployment should fail with: " +
+                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found... ");
+    }
+}

--- a/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamTracerExporterErrorTest.java
+++ b/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpUpstreamTracerExporterErrorTest.java
@@ -1,0 +1,36 @@
+package io.quarkus.opentelemetry.deployment.exporter.otlp;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class OtlpUpstreamTracerExporterErrorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withEmptyApplication()
+            .overrideConfigKey("quarkus.otel.traces.exporter", "otlp")
+            .assertException(t -> {
+                assertEquals(DeploymentException.class, t.getClass());
+                Assertions.assertTrue(t.getMessage().contains(
+                        "io.quarkus.runtime.configuration.ConfigurationException: " +
+                                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found"));
+            });
+
+    @Inject
+    OpenTelemetry openTelemetry;
+
+    @Test
+    void testOpenTelemetryButNoBatchSpanProcessor() {
+        Assertions.fail("Test should not be run as deployment should fail with: " +
+                "OpenTelemetry exporter set to 'otlp' but upstream dependencies not found... ");
+    }
+}

--- a/extensions/opentelemetry/runtime/pom.xml
+++ b/extensions/opentelemetry/runtime/pom.xml
@@ -166,6 +166,10 @@
                     <groupId>com.squareup.okhttp3</groupId>
                     <artifactId>okhttp</artifactId>
                 </exclusion>
+                <exclusion>
+                    <groupId>io.opentelemetry</groupId>
+                    <artifactId>opentelemetry-exporter-sender-okhttp</artifactId>
+                </exclusion>
             </exclusions>
         </dependency>
         <dependency>

